### PR TITLE
dag: Add `--mermaid` and `--md` option.

### PR DESCRIPTION
**Pull request by daavoo**
* [X] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

---

Now that `mermaid` [can be embedded in github md](https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/):

```bash
dvc dag --md >> dag.md
cml send-comment dag.md
```

### CODE REVIEW

**Review by skshetry — APPROVED** [review]
🔥

**Comment by skshetry** [comment]
We could extend this with `subgraph` to show individual pipelines in the future.

```
flowchart TD
	node1[data/data.xml.dvc]
	node2[evaluate]
	node3[featurize]
	node4[prepare]
	node5[train]
	node6[data2.dvc]

	subgraph P1
    node6
    end

    subgraph P2
	node1-->node4
	node3-->node2
	node3-->node5
	node4-->node3
	node5-->node2
    end
```

```mermaid
flowchart TD
	node1[data/data.xml.dvc]
	node2[evaluate]
	node3[featurize]
	node4[prepare]
	node5[train]
	node6[data2.dvc]

	subgraph P1
    node6
    end

    subgraph P2
	node1-->node4
	node3-->node2
	node3-->node5
	node4-->node3
	node5-->node2
    end
```

**Comment by daavoo** [comment]
> We could extend this with [`subgraphs`](https://mermaid-js.github.io/mermaid/#/flowchart?id=subgraphs) to show individual pipelines in the future.
> 
> ```
> flowchart TD
> 	node1[data/data.xml.dvc]
> 	node2[evaluate]
> 	node3[featurize]
> 	node4[prepare]
> 	node5[train]
> 	node6[data2.dvc]
> 
> 	subgraph P1
>     node6
>     end
> 
>     subgraph P2
> 	node1-->node4
> 	node3-->node2
> 	node3-->node5
> 	node4-->node3
> 	node5-->node2
>     end
> ```

That was a doubt I had because I couldn't find an example test. 

When it's possible for `get_pipelines` to return more than 1 pipeline? If you have many `dvc.yaml` in subfolders and call `dvc dag` on root?

**Comment by skshetry** [comment]
This is now inconsistent with other `--md` flags, in that they never add `` ``` `` (backticks).

**Comment by daavoo** [comment]
> This is now inconsistent with other `--md` flags, in that they never add `` ``` `` (backticks).

I see your point. However, the way I understand it is that `--md` outputs a self-contained MarkDown "component". 

On the one hand, the other `--md` flags output a `MarkDown table`, because it's the component that best suits the data to be shown. 

On the other, this command would generate a `MarkDown codeblock`, which is also a self-contained MarkDown component, because it's suitable for the data to be shown.

I see inconsistency in the fact that this is the first `--md` option not outputting a table, but I see consistency on all outputting self-contained MarkDown components.

**Comment by casperdcl** [comment]
Definitely prefer "self-contained markdown component with backticks"

- it's `--markdown` not `--mermaid`
- it works well with CML reports

**Comment by 0x2b3bfa0** [comment]
````console
$ dvc dag --mermaid
graph TD;
  A-->B;
  A-->C;
````
````console
$ dvc dag --mermaid --md
```mermaid
graph TD;
  A-->B;
  A-->C;
```
````

[(Source)](https://iterativeai.slack.com/archives/C700E1B3Q/p1644944835654459?thread_ts=1644920313.556279&cid=C700E1B3Q)

**Comment by 0x2b3bfa0** [comment]
I don't know what the behavior should be for `dvc dag --md` (without the `--mermaid` option) 🤔  Maybe the same as `dvc dag --mermaid --md` (?)

**Comment by casperdcl** [comment]
@daavoo as @0x2b3bfa0 said, maybe `--mermaid` (without `--md`) should not have fencing.

**Comment by 0x2b3bfa0** [comment]
````
```
🤺
```
````

**Comment by dberenbaum** [comment]
What's the status of this PR?

**Comment by daavoo** [comment]
> @daavoo as @0x2b3bfa0 said, maybe `--mermaid` (without `--md`) should not have fencing.

This is already the current status

**Review by 0x2b3bfa0 — APPROVED** [review]

File: dvc/commands/dag.py [hunk 1]
```diff
@@ -28,6 +28,32 @@ def _show_dot(G):
...
+        for node in nodes:
+            total_nodes += 1
+            node_id = f"node{total_nodes}"
+            graph += f"\n\t{node_id}[{node}]"
+            node_ids[node] = node_id
+        edges = sorted((str(a), str(b)) for b, a in pipeline.edges)
+        for a, b in edges:
+            graph += f"\n\t{node_ids[str(a)]}-->{node_ids[str(b)]}"
+    if markdown:
+        graph += "\n```"
```
Comment (skshetry) [inline dvc/commands/dag.py:53]: We can use f-string to simplify here:
```suggestion
    if markdown:
        return f"```mermaid\n{graph}\n```"
    return graph
```

File: dvc/commands/dag.py [hunk 2]
```diff
@@ -28,6 +28,32 @@ def _show_dot(G):
...
 
+def _show_mermaid(G, markdown: bool = False):
+    from dvc.repo.graph import get_pipelines
+
+    pipelines = get_pipelines(G)
+
+    graph = ""
+    if markdown:
+        graph += "```mermaid\n"
+    graph += "flowchart TD"
```
Comment (skshetry) [inline dvc/commands/dag.py:39]: If you go by the suggestion in https://github.com/iterative/dvc/pull/7383/files#r854739745.
```suggestion
    graph = "flowchart TD"
```

File: tests/unit/command/test_dag.py [hunk 3]
```diff
@@ -15,11 +23,18 @@ def test_dag(tmp_dir, dvc, mocker, fmt):
     cli_args = parse_args(args)
     assert cli_args.func == CmdDAG
 
+    if fmt == "--dot":
+        fmt_func = mocker.spy(dag, "_show_dot")
+    elif fmt in {"--mermaid", "--md"}:
+        fmt_func = mocker.spy(dag, "_show_mermaid")
```
Comment (skshetry) [inline tests/unit/command/test_dag.py:29]: Please set this outside in `pytest.parametrize`:
```python
@pytest.mark.parametrize(
    "fmt, formatted",
    [
        (None, "_show_ascii"),
        ("--dot", "_show_dot"),
        ("--mermaid", "_show_mermaid"),
        ("--md", "_show_mermaid")
    ]
)
```

File: dvc/commands/dag.py [hunk 4]
```diff
@@ -127,6 +155,18 @@ def add_parser(subparsers, parent_parser):
...
         help="Print DAG with .dot format.",
     )
+    dag_parser.add_argument(
+        "--mermaid",
+        action="store_true",
+        default=False,
+        help="Print DAG with mermaid format.",
+    )
+    dag_parser.add_argument(
+        "--md",
```
Comment (skshetry) [inline dvc/commands/dag.py:165]: Even though it's deprecated, `--show-md` flag would be nice till 3.0 for consistency. Also please set `dest=markdown` for consistency.

Here's how you can alias them:
https://github.com/iterative/dvc/blob/72d57f13bc8ad4389516ea5f22d65200799a8ac7/dvc/commands/params.py#L113-L120

You can add a test below for testing compatibility:
https://github.com/iterative/dvc/blob/72d57f13bc8ad4389516ea5f22d65200799a8ac7/tests/unit/command/test_compat_flag.py#L39

**Review by skshetry — APPROVED** [review]

**Pull request merged**